### DISCUSSION TITLE
Aligner l'applicabilité sur le littéral null

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#158] [General]
Cette refonte doit permettre de résoudre les problèmes autour de l'applicabilité et des variables manquantes :
- obligé d'ajouter des `valeur: oui` dans les espaces de noms 
- variables manquantes qui apparaissent en trop quand on évalue juste un enfant https://github.com/betagouv/publicodes/issues/30
- réponse "non" à une question qui désactive l'espace de nom de manière involontaire

---

Je me suis pas mal posé la question d'une addition ou d'un produit faisait intervenir une variable non applicable https://github.com/betagouv/publicodes/pull/66#issuecomment-878397602. En particulier ce qui me gênait était que dans la situation suivante `x` et `x + 1` était différents et donc que `0` n'était plus neutre dans l'addition :

```yaml
non_applicable:
  applicable si: non

x: non_applicable
x plus 0: x + 0

# x != x plus 0
```

Et plusieurs autres problèmes de ce genre.

Mais en fait je pense qu'on peut créer un type littéral "non applicable" qui se comporte comme le `null` de javascript :

```js
a = null         // a = null

b = null + 0     // b = 0
c = null + null  // c = 0
d = null * 0     // d = 0
e = null * null  // e = 0
```

Dans une somme ou un produit, `null` se comporte comme un zéro. Ce qui implique que si on veut uniquement connaître l'applicabilité d'une règle on n'a pas besoin de rentrer dans le détail du calcul des termes de la somme ou du produit. En revanche si on a une règle qui est directement égale à une règle non applicable elle sera elle même non applicable :

```yaml
a:
  applicable si: non

b: a # b est non applicable
```

L'idée qu'une somme est toujours applicable est ce qui a été implémenté dans #66, il faudrait le traiter pour les produits et les mécanismes dérivés (barème, etc.)

À noter que l'implémentation de #66 fonctionne avec 2 logiques distinctes lors de l’évaluation. Il serait préférable d'inférer statiquement une propriété `isNullable` (comme Typescript le fait par exemple avec le littéral `null` de JavaScript). Ainsi l'implémentation de l'applicabilité deviendrait

```js
function isApplicable(node) {
  return !(node.isNullable && evaluate(node) === null)
} 
```

**Comment by johangirod** [comment]
J'aime beaucoup le fait de distinguer `non applicable` et `non` pour la désactivation d'espace de non. Si seule les règles  `non applicable` désactive l'espace de nom, et si la situation ne peut contenir de valeur de type `non applicable`, alors il n'y a pas besoin de spécifier `valeur: oui`. 

> Il serait préférable d'inférer statiquement une propriété isNullable

Ou mieux : inférer statiquement si le type de la règle peut être `non applicable`. Cela permettrait de s'appuyer sur de l'inférence de type qui semble plus solide et définie qu'une "propriété de règle". Les règle d'inférence de type étant définies pour des mécanismes de bases (+, -, *, >, <, variations, ==). L'inférence de type des mécanismes "classiques" serait obtenue par construction : ils sont définis comme une composition de mécanisme de base.

**Reply by mquandalle** [reply]
oui quand je disais `isNullable` je pensais à de l'inférence de `non applicable`. Et effectivement sur tous les nœuds et pas seulement au niveau des règles. On parle bien de la même chose ?

**Reply by johangirod** [reply]
Ah oui, au niveau de chaque noeud, ok ! On peut commencer par cette implémentation avant d'investir dans un système d'inférence de type plus solide.

**Reply by mquandalle** [reply]
Quand je dis au niveau de chaque nœud, je pense avec inférence qui remonte jusqu'au niveau des règles :

- pour une somme et un produit `isNullable = false`
- pour une dépendance directe type `x = y` et mécanismes chaînés “unaires” (plafond, arrondi, etc.) `isNullable = valeur.isNullable`

Tu pensais à autre chose comme inférence “plus solide” ?

**Reply by johangirod** [reply]
> Tu pensais à autre chose comme inférence “plus solide” ?

De l'[inférence de type](https://en.wikipedia.org/wiki/Type_inference) généralisée à tous les types du langage. Après si la question est quelle implémentation utiliser je ne saurai te dire, j'ai pas trop regardé ce qui se faisait. Dans tous les cas, ça passerait par une passe supplémentaire sur un arbre syntaxique ne contenant que les mécanismes de base. 

Mais en attendant, on peut commencer avec une implémentation plus "artisanale"

**Comment by mquandalle** [comment]
Autre remarque, on a le cas d'usage à plusieurs endroits d'une question "parente' utilisée pour désactiver un ensemble de sous questions définies dans les règles enfants.

Avec la refacto une règle parente dont la valeur vaut "non" ne désactivera plus l'espace de nom, il faudra désormais écrire : 

```yaml
avantages . tickets resto:
  applicable si:
    nom: applicable  # peut-on faire sans ?
    question: Avez-vous le droit à des tickets resto ?
    par défaut: non 

# non applicable tant que l'on a pas précisé qu'on avait le droit à des tickets resto
avantages . tickets resto . montant du ticket:
  question: Quel montant de ticket ?
  par défaut: 5 €/ticket
```

**Reply by johangirod** [reply]
Oui, et ce qui est trop bien, c'est que la valeur de l'exonération pourra remonter dans la règle parente : plus besoin de définir à chaque fois une règle "montant" ! 

Par contre, il y aura un peu de tuyauterie à faire pour que la question soit reliée à ticket restau au niveau des infobulles d'aide. Mais rien de très méchant. 

C'est clairement un breaking, mais ça me semble aller dans une bonne direction.

**Reply by mquandalle** [reply]
Une alternative est d'exposer le littéral null en publicodes, par exemple `x: [non applicable]` et de l'autoriser comme possibilité de réponse.

Mais je crois que je préfère l'imbrication de la question dans `applicable si`.

**Reply by mquandalle** [reply]
Vu ensemble avec @johangirod, on préfère remonter les règles de type `. montant` au niveau de l'espace de nom et déplacer les questions « d'activation » dans le `applicable si`.

Cette réorganisation des règles implique effectivement de devoir nommer explicitement toutes les questions, pas seulement avec le nom de l'espace mais avec un nom dédié pour chaque question. Après un rapide survol de la base de règles mon-entreprise, je pense que c'est une évolution plutôt positive, à tester.

**Comment by mquandalle** [comment]
### Parent rendu non applicable sauf dans un enfant

Actuellement l'implémentation de la désactivation de branche fonctionne en intégrant une référence vers le parent dans l'AST de chaque règle. Cette référence peut ensuite être modifiée par des remplacements / rend non applicable.

Ça veut dire qu'il est théoriquement possible de faire la chose suivante :


```yaml
parent:
parent . enfant:  5
x:
  remplace:
    - règle: parent
      par: non
      sauf dans: parent . enfant
```

À noter que ça ne fonctionne que pour le premier parent, ceux d'au dessus ne sont pas référencés dans l'AST.

J'envisage de retirer les références au parent de l'AST car elles constituent des cycles (je crois aussi qu'un AST doit rester proche de la syntaxe du programme saisie, ce qui était un peu mon désaccord sur certaines transformations en mécanismes de base à ce state là de l’exécution).

Ça veut dire que le parent rendu non applicable sauf dans un enfant ne serait plus possible. Ça me semble plutôt raisonnable, je pense qu'on a pas besoin de cette fonctionnalité. D'autant que si on imagine le remplacement comme une substitution dans le code source de la règle, le fait que la règle parente ne soit pas explicitement présente dans la formule de calcul de l'enfant plaide aussi en faveur de ne pas l'intégrer dans le champs des remplacements. (j'ai essayé de l'intégrer et ça complique pas mal l'implémentation).

**Reply by johangirod** [reply]
Ça me paraît complètement raisonnable, et j'irai même jusqu'à affirmer que le comportement actuel n'a pas été voulu et ne devrait pas être.

**Comment by mquandalle** [comment]
### Changement cassant : types littéraux null et undefined

La PR associée https://github.com/betagouv/publicodes/pull/159 change les types littéraux utilisés dans publicodes et exposés dans les `nodeValue` retournées.

. | Avant | Après
--- | --- | ---
Règle non définie | `null` | `undefined`
Règle non applicable | `false` | `null`

Ce changement cassant n'était pas strictement nécessaire mais il vise à mieux aligner la sémantique de ces opérations avec le comportant Javascript :

- `false` pour non applicable posait un problème de confusion avec `false` comme réponse booléenne à une question
- le type `undefined` de Javascript se “propage en erreurs” dans les opérations, c'est à dire que `undefined + 4` est une erreur, ce qui est aussi le comportement implémenté lorsqu'on rencontre une règle qui n'a pas de définition
- le type `null` de Javascript se comporte comme un zéro dans les opération, par exemple `null + 4 = 4` ou `null * null = 0`, c'est également le cas d'une règle non applicable (même si les mécanismes peuvent lui donner une valeur par défaut, par exemple un plafond `null` vaut `Infinity` et non pas 0, mais c'est assez cohérent avec l'utilisation du littéral `null` dans ce contexte).

**Comment by johangirod** [comment]
#373 pousse plus en profondeur les réflexions sur l'applicabilité des espaces de nom
